src/Language/Haskell/TH/Datatype.hs

```diff
@@ -1192,20 +1278,56 @@ typeToPred t =
 #endif
 
 -- | Decompose a type into a list of it's outermost applications. This process
--- forgets about infix application and explicit parentheses.
+-- forgets about infix application, explicit parentheses, and visible kind
+-- applications.
 --
 -- This operation should be used after all 'UInfixT' cases have been resolved
 -- by 'resolveFixities' if the argument is being user generated.
 --
 -- > t ~= foldl1 AppT (decomposeType t)
 decomposeType :: Type -> NonEmpty Type
```

Comment:
I wonder if I should make this return type `(Type, [Type])` to be consistent with `decompseTypeArgs`? (We don't export `decomposeType` at the moment, so we could do so without breaking anything API-wise.)

Reply:
I decided to punt on this for now. (If we do revisit this, perhaps we should also remove `NonEmptySnoc` and replace it with `([a], a)`.)